sdks/java/io/amazon-web-services2/src/test/java/org/apache/beam/sdk/io/aws2/schemas/AwsSchemaProviderTest.java

```diff
@@ -0,0 +1,269 @@
+                    attribute(b -> b.binaryValue(sdkBytes("v")).dataType("Binary")),
+                    "stringList",
+                    attribute(b -> b.stringListValues("v1", "v2")),
+                    "binaryList",
+                    attribute(b -> b.binaryListValues(sdkBytes("v1"), sdkBytes("v2")))))
+            .build();
+
+    SchemaCoder<SendMessageRequest> coder = registry.getSchemaCoder(SendMessageRequest.class);
+
+    Row row = coder.getToRowFunction().apply(request);
+    assertThat(coder.getFromRowFunction().apply(row)).isEqualTo(request);
+
+    byte[] requestBytes = encodeToByteArray(coder, request);
+    SendMessageRequest requestFromBytes = decodeFromByteArray(coder, requestBytes);
+    assertThat(requestFromBytes).isEqualTo(request);
+
+    // verify still serializable after use
+    ensureSerializableRoundTrip(coder.getToRowFunction());
+    ensureSerializableRoundTrip(coder.getFromRowFunction());
+  }
+
+  @Test
+  public void testFromRowWithPartialSchema() throws NoSuchSchemaException {
+    SerializableFunction<Row, SendMessageRequest> fromRow =
+        registry.getFromRowFunction(SendMessageRequest.class);
+
+    Schema partialSchema =
+        Schema.builder()
+            .addNullableField("queueUrl", FieldType.STRING)
+            .addNullableField("messageBody", FieldType.STRING)
+            .addNullableField("delaySeconds", FieldType.INT32)
+            .build();
+
+    SendMessageRequest request =
+        SendMessageRequest.builder()
+            .queueUrl("queue")
+            .messageBody("body")
+            .delaySeconds(100)
+            .build();
+
+    Row row = Row.withSchema(partialSchema).addValues("queue", "body", 100).build();
+
+    assertThat(fromRow.apply(row)).isEqualTo(request);
+  }
+
+  @Test
+  public void testFailFromRowOnUnknownField() throws NoSuchSchemaException {
+    SerializableFunction<Row, SendMessageRequest> fromRow =
+        registry.getFromRowFunction(SendMessageRequest.class);
+
+    Schema partialSchema =
+        Schema.builder()
+            .addNullableField("queueUrl", FieldType.STRING)
+            .addNullableField("messageBody", FieldType.STRING)
+            .addNullableField("unknownField", FieldType.INT32)
+            .build();
+
+    Row row = Row.withSchema(partialSchema).addValues("queue", "body", 100).build();
+
+    assertThatThrownBy(() -> fromRow.apply(row))
+        .isInstanceOf(IllegalStateException.class)
+        .hasMessageContaining("Row schema contains unknown fields: [unknownField]");
+  }
+
+  private static MessageAttributeValue attribute(
+      Function<MessageAttributeValue.Builder, MessageAttributeValue.Builder> b) {
+    return b.apply(MessageAttributeValue.builder()).build();
+  }
+
+  private static SdkBytes sdkBytes(String str) {
+    return SdkBytes.fromByteArrayUnsafe(bytes(str));
+  }
+
+  private static byte[] bytes(String str) {
+    return str.getBytes(UTF_8);
+  }
+
+  private <T> Condition<Row> field(String name, T value) {
```

Comment:
This is handy, maybe it should be in [SchemaTestUtils](https://github.com/apache/beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/schemas/utils/SchemaTestUtils.java) for discoverability?

Reply:
This code depends on `assertj-core` and probably doesn't fit into SchemaTestUtils for that reason.
I could rewrite it as `Matcher` if you think it's worth it,e.g.:
```java
public static Matcher<Row> containsFieldValues(Map<String, Object> expectedFieldValues) 
```